File: leetcode/chain/24.py
# Definition for singly-linked list.
class ListNode:

    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:

    def swapPairs(self, head: ListNode) -> ListNode:
        '''两两交换链表中的节点
        
        @Note:
            新增虚拟节点
        '''
        ret = ListNode()
        ret.next = head
        last = ret
        while head is not None:
            tmp = head.next
            if tmp is not None:
                head.next = tmp.next
                tmp.next = head
                last.next = tmp
                last = head  # 挪位
                head = head.next  #挪位
            else:
                break
        return ret.next

File: leetcode/chain/24.cpp
#include <iostream>

#include "list_node.h"

class Solution {
 public:
  ListNode* swap_pairs(ListNode* node) {
    ListNode *ret = new ListNode(0, node);
    ListNode *last = ret;
    while (node != nullptr) {
      ListNode *next = node->next;
      if (next != nullptr) {
        node->next = next->next;
        next->next = node;
        last->next = next;

        last = node;  // 挪位
        node = node->next; // 挪位
      } else {
        break;
      }
    }
    return ret->next;
  }
};

int main() {
  ListNode* node = new ListNode(3);
  node->append(new ListNode(4));
  node->append(new ListNode(4));
  node->append(new ListNode(5));
  node->print();

  Solution solution = Solution();
  ListNode* ret = solution.swap_pairs(node);
  ret->print();
}